src/ucp/tag/tag_recv.c

```diff
@@ -151,9 +147,63 @@ ucs_status_ptr_t ucp_tag_recv_nb(ucp_worker_h worker, void *buffer, size_t count
         req->recv.tag      = tag;
         req->recv.tag_mask = tag_mask;
         ucs_queue_push(&worker->context->tag.expected, &req->recv.queue);
-        ucs_trace_req("recv_nb returning expected request %p (%p)", req, req + 1);
+        ucs_trace_req("recv_nb%c returning expected request %p (%p)",
+                      (req->flags & UCP_REQUEST_FLAG_EXTERNAL) ? 'r' : ' ',
+                      req, req + 1);
     }
 
+    return status;
+}
+
+ucs_status_t ucp_tag_recv_nbr(ucp_worker_h worker, void *buffer, size_t count,
+                              uintptr_t datatype, ucp_tag_t tag,
+                              ucp_tag_t tag_mask, void *request)
+{
+    ucp_request_t  *req = (ucp_request_t *)request - 1;
+    ucs_status_t status;
+
+    ucp_tag_recv_request_init(req, worker, buffer, count, datatype);
+
+    UCS_INSTRUMENT_RECORD(UCS_INSTRUMENT_TYPE_UCP_RX, "ucp_tag_recv_nbr",
+                          req, ucp_contig_dt_length(datatype, count));
+
+    req->flags   |= UCP_REQUEST_FLAG_EXTERNAL;
```

Comment:
maybe could pass "extra_flags" to ucp_tag_recv_request_init() to avoid this additional |=